Add OpenCover PS Module to collect code coverage

adityapatwardhan opened the pull request:
OpenCover PS Module helps is collecting Code Coverage using the OpenCover
toolset. The module helps in comparing two code coverage runs as well. 

Contributions from @JamesWTruher  as well.

adityapatwardhan committed:
Add OpenCover PS Module to collect code coverage

OpenCover PS Module helps is collecting Code Coverage using the OpenCover
toolset. The module helps in comparing two code coverage runs as well.

msftclas commented:
Hi __@adityapatwardhan__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Aditya Patwardhan). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

lzybkr commented:
git thinks test/tools/OpenCover/OpenCover.psd1 is a binary file - can you make sure the file is ascii?

JamesWTruher commented:
>  A u t h o r   =   ' a d i t y a p ' 

not sure if you want your alias to leak - perhaps git username?

---
Refers to: test/tools/OpenCover/OpenCover.psd1:47 in dffa990. [](commit_id = dffa99064f12ddd25bb08c8c7a9f3a32decb60da, deletion_comment = False)

JamesWTruher commented:
> ��# 

looks like unicode - perhaps just ascii

---
Refers to: test/tools/OpenCover/OpenCover.psd1:1 in dffa990. [](commit_id = dffa99064f12ddd25bb08c8c7a9f3a32decb60da, deletion_comment = False)

adityapatwardhan committed:
Change OpenCover.psd1 to ASCII

adityapatwardhan commented:
@lzybkr @JamesWTruher I have pushed a change to address comments.

JamesWTruher commented:
:shipit:

adityapatwardhan committed:
Fix an error in path for OpenCover

Fixed an error on path for OpenCover. Also used ZipFile class instead of
cmdlet as it might not be available on CI system.

adityapatwardhan commented:
Please hold off the review. I will update this module so that it supports Powershell v4.0.

adityapatwardhan committed:
Convert module to be Powershell v4 compliant

Changed implementation from classes to PSObjects and implemented
Expand-ZipArchive.

adityapatwardhan commented:
@lzybkr @JamesWTruher The module is ready for review now.

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+
+function Get-AssemblyCoverageData([xml.xmlelement] $element)
+{
+    $AssemblyCoverageData = New-Object System.Management.Automation.PSObject
```
The following would be much faster:
```powershell
$coverageSummary = Get-CoverageSummary -element $element.Summary
$assemblyCoverageData = [pscustomobject]@{
    AssemblyName = $element.ModuleName
    Branch = $coverageSummary.BranchCoverage
    Sequence = $coverageSummary.SequenceCoverage
    CoverageSummary = $coverageSummary
}
$AssemblyCoverageData | Add-Member -MemberType ScriptMethod -Name ToString -Value { "{0} ({1})" -f $this.AssemblyName,$this.CoverageSummary.BranchCoverage } -Force
return $AssemblyCoverageData
```

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
-        $this.Sequence = $this.Coverage.SequenceCoverage
+function Get-CodeCoverageChange($r1, $r2)
+{
+    $CoverageChange = New-Object System.Management.Automation.PSObject
```
Same here - use the [pscustomobject] cast.

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+function Get-AssemblyCoverageData([xml.xmlelement] $element)
+{
+    $AssemblyCoverageData = New-Object System.Management.Automation.PSObject
+    $AssemblyCoverageData | Add-Member -MemberType NoteProperty -Name AssemblyName -TypeName [string] -Value ([string]::Empty)
```
The `-TypeName` isn't working how you think here - it's adding the literal string `[string]` to PSTypeNames on `$AssemblyCoverageData.

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+
+    if ( compare-object $r1.assemblyname $r2.assemblyname ) { throw "different assemblies" }
+
+    $AssemblyCoverageChange = New-Object System.Management.Automation.PSObject
```
Again, use `[pscustomobject]@{...}`

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+    [xml]$CoverageXml = get-content -readcount 0 $xmlPath
+    if ( $CoverageXml.CoverageSession -eq $null ) { throw "CoverageSession data not found" }        
+
+    $CoverageData = New-Object System.Management.Automation.PSObject
```
Again, use `[pscustomobject]@{ ... }`

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+
+    $CoverageData.CoverageSummary = (Get-CoverageSummary -element $CoverageXml.CoverageSession.Summary)
+
+    remove-variable CoverageXml
```
Assign to $null is sufficient to allow garbage collection.

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+    $CoverageData.CoverageSummary = (Get-CoverageSummary -element $CoverageXml.CoverageSession.Summary)
+
+    remove-variable CoverageXml
+    [gc]::Collect()
```
I'm sure you know this - but explicit calls to garbage collection are normally discouraged.
If you need to keep it for some reason - there should be a comment justifying the code.

JamesWTruher replied:
This object is *very* large - larger than 1GB, which will relegate it to Gen2 storage (ala LOH) which is not collected often - I saw more than 10/15 minutes elapse after assignment to $null and still no clean up,
a comment as to why it's being done explicitly would certainly not go amiss

---
In reply to: [86645728](https://github.com/PowerShell/PowerShell/pull/2585#discussion_r86645728) [](ancestors = 86645728)

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -1,113 +1,187 @@
+
+function Get-CoverageSummary([xml.xmlelement] $element)
+{
+    $CoverageSummary = New-Object System.Management.Automation.PSObject
```
Again, `[pscustomobject]@{ ... }`

adityapatwardhan committed:
Added CodeCoverage as a configuration to project.json files

Added CodeCoverage as the new configuration for all the project.json
files. When Start-PSBuild is executed with configuration as CodeCoverage,
we change the degubType to 'full' as required by OpenCover toolset.

Also made changes to appveyor.psm1 to build a CodeCoverage package on
daily builds and publish it as a zip.

adityapatwardhan commented:
@lzybkr @JamesWTruher @andschwa This PR is ready for review again. I have added CodeCoverage as an configuration for Start-PSBuild. This is done as we need debugType as Full for OpenCover.

Appveyor.psm1 is updated for generating daily builds for CodeCoverage configuration and publishing them as a zip file.

JamesWTruher commented on src/Microsoft.PackageManagement.NuGetProvider/project.json:
```diff
@@ -9,9 +9,25 @@
         "Linux": {
             "buildOptions": {
-                "define": ["UNIX"]
+                "define": [ "UNIX" ],
```
>  [](start = 27, length = 1)

i'm assuming that the extra spaces in this are harmless

adityapatwardhan committed:
Addressed code review comments

Changed from Add-Member to use pscustomobject type accelator. Removed
[gc]::collect.

adityapatwardhan committed:
Added explicit garbage collection

JamesWTruher commented:
:shipit:

adityapatwardhan commented:
@lzybkr @andschwa Updated with addressing code review comments.

andschwa commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -9,9 +9,25 @@
             "buildOptions": {
-                "define": ["UNIX"]
+                "define": [ "UNIX" ],
+                "debugType": "portable"
```
Since this is the default, it probably does not need to be included.

adityapatwardhan replied:
@andschwa, it is a default only for Linux right?

andschwa commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -9,9 +9,25 @@
+                "debugType": "portable"
+            }
+        },
+        "Debug": {
```
These don't need to be added; they're default .NET CLI configurations.

andschwa commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -44,8 +60,7 @@
             "imports": ["dnxcore50"],
             "buildOptions": {
-                "define": ["CORECLR"],
-                "debugType": "portable"
```
Hm... I don't know why this was here. I was pretty sure it's the default, but it was purposefully added in https://github.com/PowerShell/PowerShell/pull/1959.

OH. [It defaults to `full` on Windows](https://github.com/dotnet/cli/blob/04f40f906dce2678d80fb9787e68de76ee6bf57e/src/dotnet/commands/dotnet-compile-csc/Program.cs#L201), and the previous PR overrode it.

andschwa commented:
/cc @daxian-dbw

lzybkr commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -0,0 +1,418 @@
+    param ( 
+        [parameter()]$OutputLog = "$pwd/OpenCover.xml",
+        [parameter(Mandatory=$true)]$TestDirectory,
+        [parameter()]$OpenCoverPath = "$HOME\AppData\Local\Apps\OpenCover", # this is the default install location for open cover
```
How is `$env:LOCALAPPDATA` (and you should use that instead of `$HOME\AppData\Local`) the default? I tried `Install-OpenCover` and it used `$pwd`.

adityapatwardhan replied:
AppData\Local\.. is the default only if it is installed through MSI. Changed it a mandatory parameter and removed the default.

andyleejordan reviewed (approved):
To summarize the `project.json` changes:

This makes `debugType` continue to be `portable` on Windows (which, according to .NET CLI source code, is the default on Linux, but `full` is the default on Windows) for the Debug and Release configurations, where the `CodeCoverage` configuration explicitly overrides it to `full`.

It needs to be `portable` for VS Code debugging (per #1959), and since it had to be removed from the `framework` section, the default `Debug` and `Release` configurations had to be explicitly added.

I'd appreciate this explanation being included in the commit messages, as I had to dig around to find out why the changes were made in this specific way, but I sign off on the changes.

adityapatwardhan commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -0,0 +1,418 @@
+    $webclient = [System.Net.WebClient]::New()
+    $filename =  "opencover.${version}.zip"
+    $packageUrl = "https://ci.appveyor.com/api/buildjobs/xj78v6dac42uob8q/artifacts/main%2Fbin%2Fzip%2Fopencover.4.6.589.zip"
+    if ( test-path $PWD/$Filename ) 
```
Changed to $env:TEMP

adityapatwardhan commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -0,0 +1,418 @@
+        }
+        else 
+        {
+            throw "package already exists, not downloading"
```
Added the path where the package was found to the error message. Made targetDirectory mandatory.

daxian-dbw commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -9,9 +9,25 @@
+        "CodeCoverage": {
+                "buildOptions": {
+                    "debugType": "full"
+                }
```
The changes in project.json files are respecting the existing indentation pattern. Could you please update those changes to use the same existing indentation patterns?

daxian-dbw commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -44,8 +60,7 @@
             "buildOptions": {
-                "define": ["CORECLR"],
-                "debugType": "portable"
+                "define": ["CORECLR"]
```
A question, before this change, the `.pdb` produced for CoreCLR are `portable`, but this setting doesn't affect building target `.net451`. After the change, I guess the setting `"debugType": "portable"` will also apply to full clr building, right? If so, what does that imply? Does the portable `pdb` work for FullCLR powershell?

daxian-dbw commented on tools/appveyor.psm1:
```diff
@@ -159,6 +159,10 @@ function Invoke-AppVeyorBuild
+      if(Test-DailyBuild)
+      {
+          Start-PSBuild -CrossGen -Configuration 'CodeCoverage'
+      }
```
Building `-Configuration 'CodeCoverage'` will produce wrong packages with `Start-PSPackage` in the `finish` stage. This is because `Start-PSPackage` assumes the more recent build is targeting .net core and with `release` configuration, see [here](https://github.com/PowerShell/PowerShell/blob/master/build.psm1#L1126). You can move this if block before `Start-PSBuild -CrossGen -Configuration $buildConfiguration` or `Start-PSBuild -FullCLR` to work around it.

Moreover, [this line](https://github.com/PowerShell/PowerShell/blob/master/build.psm1#L1102) should be changed to `$Script:Options.Configuration -ne $Configuration -or` since `Configuration` takes new value now.

adityapatwardhan committed:
Addressed code review comments

- Make sure that the build Start-PSPackage gets is not a code coverage
  build
- Add debugType = full for FullCLR
- Remove configurations from PackageManagement files as it is not needed.
- Build CodeCoverage build first in AppVeyor.

adityapatwardhan committed:
Resolve merge conflict

adityapatwardhan commented:
@daxian-dbw @lzybkr Addressed code review comments and merge conflict.

daxian-dbw commented on build.psm1:
```diff
@@ -1096,10 +1102,11 @@ function Start-PSPackage {
+    if (-not $Script:Options -or                                ## Start-PSBuild hasn't been executed yet
+        -not $Script:Options.CrossGen -or                       ## Last build didn't specify -CrossGen
+        $Script:Options.Runtime -ne $Runtime -or                ## Last build wasn't for the required RID
+        $Script:Options.Configuration -eq "Debug" -or           ## Last build was with 'Debug' configuration
```
> $Script:Options.Configuration -eq "Debug" -or

This line can be removed.

daxian-dbw commented on src/Microsoft.PackageManagement.ArchiverProviders/project.json:
```diff
@@ -49,4 +49,4 @@
         }
     }
-}
+}
```
unnecessary end of line changes probably should be avoided.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Diagnostics/project.json:
```diff
@@ -10,6 +10,30 @@
+        "Debug": {
+                "buildOptions": {
+                    "debugType": "portable"
+                }
```
You can see the `buildOptions` section is still not following the existing indentation patterns. The existing indentation pattern is to indent 4 spaces for any enclosing sections. Many other changes in project.json files also have the same issue. Could you please fix them all? Thanks!

adityapatwardhan replied:
@daxian-dbw Fixed. Sorry missed the buildOptions. Had only fixe the closing braces. Now it is fixed.

daxian-dbw replied:
Thanks!

adityapatwardhan committed:
Fix indentation

adityapatwardhan committed:
Fix newline at end of file

adityapatwardhan commented:
@daxian-dbw Fixed end of file newline issue.

adityapatwardhan commented:
@lzybkr Can you have a look? I have addressed your concerns.

JamesWTruher commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -0,0 +1,418 @@
+
+    # check to be sure that OpenCover is present
+    $OpenCoverBin = "$OpenCoverPath\opencover.console.exe"
+    if ( ! (test-path $OpenCoverBin)) 
```
@lzybkr your suggestion requires that the open cover binaries must be installed into a location covered by $env:PATH which should not be needed or even desired

adityapatwardhan committed:
Added command discovery for locating OpenCover.console.exe

adityapatwardhan commented:
@lzybkr Added command lookup as suggested. Please have a look.
